hw_31_snake/index.js

```diff
@@ -0,0 +1,255 @@
+    static snakeCellCssClass = 'snake-cell';
+    static snakeCssClass = 'snake';
+    static snakeHeadCssClass = 'snake-head';
+    static snakeBodyCssClass = 'snake-body';
+    static gridContainerSelector = '#snake-container';
+
+    #snake = [];
+    #process = null;
+    #speed = 0;
+    #startBtn = this.find('#snake-start-game');
+    #endBtn = this.find('#snake-end-game');
+    #form = this.find('#snake-controls-form');
+    #messageContainer = this.find('#snake-message');
+    #scoreContainer = this.find('#snake-score');
+    #score = 0;
+    #foodCoords = null;
+    #foodContainer = this.#findByCoords({ cell: 1, row: 1 });
+
+
+
+    constructor({ boxSize, gridCount, scoreToWin }) {
+        super({
+            boxSize,
+            gridCount,
+            gridCellCssClass: Snake.snakeCellCssClass,
+            gridContainerSelector: Snake.gridContainerSelector,
+        });
+        this.direction = D.LEFT;
+        this.scoreToWin = scoreToWin;
+
+        this.#init();
+    }
+
+
+    #init() {
+        this.#startBtn.addEventListener('click', () => {
+            this.#start();
+        });
+
+        this.#endBtn.addEventListener('click', () => {
+            this.#endGame();
+        });
+
+        document.addEventListener('keydown', (event) => {
+            this.#updateDirection(event);
+        })
+    }
+
+    #start() {
+        this.#score = 0;
+        this.#scoreContainer.firstElementChild.textContent = `${this.#score}`;
+        let middleCell = Math.floor(this.gridCount) / 2;
+        this.#snake = this.#buildSnake(middleCell, middleCell);
+        this.#speed = +this.#form.speed.value;
+       
+        this.#startBtn.style.display = "none";
+        this.#endBtn.style.display = "block";
+        this.#messageContainer.innerHTML = "Welcome to Snake!";
+
+        this.#process = setInterval(() => {
+
+            let { cell, row } = this.#noWallMode();
+            
+            let snakePartToShift = null;
+
+            switch (this.direction) {
+                case D.LEFT: {
+                    snakePartToShift = {
+                        cell: cell - 1,
+                        row,
+                    };
+                }; break;
+                case D.RIGHT: {
+                    snakePartToShift = {
+                        cell: cell + 1,
+                        row,
+                    };
+                }; break;
+                case D.UP: {
+                    snakePartToShift = {
+                        cell,
+                        row: row - 1,
+                    };
+                }; break;
+                case D.DOWN: {
+                    snakePartToShift = {
+                        cell,
+                        row: row + 1,
+                    };
+                }; break;
+            };
+
+            this.#snake.unshift(snakePartToShift);
+
+            this.#clear();
+            if (this.#checkTailCrash() || (this.#checkIfFoodEaten() && this.#score === this.scoreToWin)) {
```

Comment:
It's better to call endGame inside the `this.#checkTailCrash`. 
`this.#checkIfFoodEaten()` does not affect the end of the game.

The `this.#score === this.scoreToWin` expression already used in endGame